types/react/index.d.ts

```diff
@@ -90,6 +90,42 @@ declare namespace React {
     type RefCallback<T> = { bivarianceHack(instance: T | null): void }["bivarianceHack"];
     type Ref<T> = RefCallback<T> | RefObject<T> | null;
     type LegacyRef<T> = string | Ref<T>;
+    /**
+     * Gets the instance type for a React element. The instance will be different for various component types:
+     *
+     * - React class components will be the class instance. So if you had `class Foo extends React.Component<{}> {}`
+     *   and used `React.ElementRef<typeof Foo>` then the type would be the instance of `Foo`.
+     * - React stateless functional components do not have a backing instance and so `React.ElementRef<typeof Bar>`
+     *   (when `Bar` is `function Bar() {}`) will give you the `undefined` type.
+     * - JSX intrinsics like `div` will give you their DOM instance. For `React.ElementRef<'div'>` that would be
+     *   `HTMLDivElement`. For `React.ElementRef<'input'>` that would be `HTMLInputElement`.
+     * - React stateless functional components that forward a `ref` will give you the `ElementRef` of the forwarded
+     *   to component.
+     *
+     * `C` must be the type _of_ a React component so you need to use typeof as in React.ElementRef<typeof MyComponent>.
+     * 
+     * @todo In Flow, this works a little different with forwarded refs and the `AbstractComponent` that
+     *       `React.forwardRef()` returns.
+     */
+    type ElementRef<
+        C extends
+            | { new (props: any): Component<any> }
+            | (() => JSX.Element)
+            | keyof JSX.IntrinsicElements
+            | ForwardRefExoticComponent<any>
```

Comment:
@ferdaber Would you prefer `ExoticComponent<any>` here?

Reply:
Well, this is specifically about components that forward refs, not just any exotic component, so I’m guessing this is fine 🤔

Reply:
@jbrown215 Hmm, given your comment I suspect this should be `ExoticComponent` after all–which is defined [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/27123c07449f1fac56c5ec735737e29bff57dac4/types/react/index.d.ts#L310-L326).

But perhaps I misunderstood what `memo`, `Fragment`, and `Suspense` etc return. (I’ll have to take a look tomorrow.)

Reply:
Does a memoized forwardRef component still forwards its ref? This is a run-time question

Reply:
I was just looking into this. Without validating it myself, I see at least two cases that says `memo` does, one of which is on the `flow` repo:

* https://github.com/facebook/flow/issues/7815
* https://github.com/yannickcr/eslint-plugin-react/issues/2349

In any case, I already added a type test-case for it and it already worked as expected.

Will look at `lazy` now.

Reply:
Added a test-case for `lazy`, which also worked out of the box.